File: codebase/superdarn/src.java/fitacfdisplay/Time.java
/* Time.java
   =========
   Author R.J.Barnes
*/

/*
 $License$
*/




public class Time {

  int DAY_SEC=(24*3600); 
  int YEAR_SEC=(365*24*3600);
  int LYEAR_SEC=(366*24*3600);

  int year;
  int month;
  int day;
  int hour;
  int minute;
  double second;

  public void setYrSec(int year,int time) {
    int yd;
    int dt;
    int n;
    int nday[]={0,31,59,90,120,151,181,212,243,273,304,334};
    int lday[]={0,31,60,91,121,152,182,213,244,274,305,335};
    int jday[];

    if ((year % 4)==0) jday=lday;
      else jday=nday;

    yd=time/(24*60*60);
    for (n=0;(n<12) && (yd>=jday[n]);n++);
   
    this.year=year; 
    this.month=n;
    
    if (n>0) this.day=1+yd-jday[n-1];
    else this.day=yd+1;
     
    dt=time % (24*60*60);
    this.hour=dt/(60*60);
    this.minute=(dt % (60*60))/60;
    this.second=(double) (dt % 60);
  }

  public int getYrSec() {

    int t;
    int jday[] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334};
    int mday[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};

    if ( (this.month < 1) || (this.month > 12) || 
         (this.hour < 0) ||  (this.hour > 23) || 
         (this.day <1) || (this.minute < 0) || 
         (this.minute > 59) || (this.second < 0) || 
         (this.second >= 60) )
	   return -1;

    if (this.day > mday[this.month-1])
      if (this.month != 2) return -1;
      else if (day != (mday[1] +1) || ((this.year % 4) != 0)) return -1;

    t = jday[this.month-1] + this.day - 1;

    if ( (this.month > 2) && ((this.year % 4) == 0)) t = t+1;

    return (((t*24 + this.hour)*60 + this.minute)*60)+(int) this.second;
  }


  double getEpoch() {
    double tme;
    int lpyear,ryear;
    int yrsec;
    if (this.year<1970) return -1;
    yrsec=getYrSec();
    if (yrsec==-1) return -1;
    tme=yrsec+(this.second-(int) this.second);
    
    /* work out number of seconds from 1970 to start of year */

    lpyear=(this.year-1969)/4;
    ryear=(this.year-1970)-lpyear;

    tme+=(lpyear*LYEAR_SEC)+(ryear*YEAR_SEC); 
    return tme;
  }


  void setEpoch(double tme) {
    int yrsec=0;
    int sec;
    int i=0;
    int yr;

    /* find the year */

    while ((yrsec<=tme) && (i<1e6)) {
      if ((i % 4)==2) yrsec+=LYEAR_SEC;
      else yrsec+=YEAR_SEC;
      i++;
    }
    if (((i-1) % 4)==2) tme-=yrsec-LYEAR_SEC;
    else tme-=yrsec-YEAR_SEC;
    yr=i+1969;
    setYrSec(yr,(int) tme);

    this.second=this.second+(tme-(int) tme);
  }
 
}


   

File: codebase/general/src.lib/time.1.7/src/time.c
/* time.c
   ======
   Author: R.J.Barnes
 Copyright (c) 2012 The Johns Hopkins University/Applied Physics Laboratory

This file is part of the Radar Software Toolkit (RST).

RST is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

Modifications:

2021-08-27 Angeline G Burrell: Moved string-to-epoch-time routines here from
           make_grid.2.0 to allow better access across all tools.

*/
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include "rtime.h"


#define DAY_SEC 86400

int TimeYMDHMSToYrsec(int yr,int mo,int dy,int hr,int mn,int sec) {

  time_t clock;
  struct tm tm;
  char *tz;

  memset(&tm,0,sizeof(struct tm));
  tm.tm_year=yr-1900;
  tm.tm_mon=0;
  tm.tm_mday=1;
  tm.tm_hour=0;
  tm.tm_min=0;
  tm.tm_sec=0;

  tz = getenv("TZ");
  setenv("TZ", "", 1);
  tzset();
  clock=mktime(&tm);

  memset(&tm,0,sizeof(struct tm));
  tm.tm_year=yr-1900;
  tm.tm_mon=mo-1;
  tm.tm_mday=dy;
  tm.tm_hour=hr;
  tm.tm_min=mn;
  tm.tm_sec=sec;
  clock=mktime(&tm)-clock;
  

  if (tz) setenv("TZ", tz, 1);
  else unsetenv("TZ");
  tzset();

  return (int) clock;
}

void TimeYrsecToYMDHMS(int yrsec,int yr,int *mo,int *dy,int *hr,int *mn,
	                int *sc) {


  time_t clock;
  struct tm tmyr;
  struct tm *tm;
  char *tz;

  memset(&tmyr,0,sizeof(struct tm));
  tmyr.tm_year=yr-1900;
  tmyr.tm_mon=0;
  tmyr.tm_mday=1;
  tmyr.tm_hour=0;
  tmyr.tm_min=0;
  tmyr.tm_sec=0;

  tz = getenv("TZ");
  setenv("TZ", "", 1);
  tzset();
 
  clock=mktime(&tmyr);
  if (tz) setenv("TZ", tz, 1);
  else unsetenv("TZ");
  tzset();

  clock=clock+yrsec;
  tm=gmtime(&clock);
  
  *mo=tm->tm_mon+1;
  *dy=tm->tm_mday;
  *hr=tm->tm_hour; 
  *mn=tm->tm_min;
  *sc=tm->tm_sec;
}


double TimeYMDHMSToEpoch(int yr,int mo,int dy,int hr,int mn,double sec) {
 
  time_t clock;
  struct tm tm;
  char *tz;

  memset(&tm,0,sizeof(struct tm));
  tm.tm_year=yr-1900;
  tm.tm_mon=mo-1;
  tm.tm_mday=dy;
  tm.tm_hour=hr;
  tm.tm_min=mn;
  tm.tm_sec=floor(sec);

  tz = getenv("TZ");
  setenv("TZ", "", 1);
  tzset();
  clock=mktime(&tm);
  if (tz) setenv("TZ", tz, 1);
  else unsetenv("TZ");
  tzset();
               
  return clock+(sec-floor(sec));
}

void TimeEpochToYMDHMS(double tme,int *yr,int *mo,int *dy,int *hr,int *mn,
	       double *sc) {
  time_t clock;
  struct tm *tm;

  clock=floor(tme);
  tm=gmtime(&clock);
  
  *yr=tm->tm_year+1900;
  *mo=tm->tm_mon+1;
  *dy=tm->tm_mday;
  *hr=tm->tm_hour; 
  *mn=tm->tm_min;
  *sc=tm->tm_sec+(tme-floor(tme));
}

double TimeYMDHMSToJulian(int yr,int mo,int dy,int hr,int mt,double sc) {

  int A,B,i;
  double jdoy;
  double dfrac;
  yr=yr-1;
  i=yr/100;
  A=i;
  i=A/4;
  B=2-A+i;
  i=365.25*yr;
  i+=30.6001*14;
  jdoy=i+1720994.5+B;

  
  dfrac=1.0+(double)TimeYMDHMSToYrsec(yr+1,mo,dy,hr,mt,sc)/DAY_SEC;
   
  return jdoy+dfrac; 

}


int TimeJulianToYMDHMS(double jd,int *yr,int *mo,
                 int *dy,int *hr,int *mt,double *sc) {

  int Z,month;
  int hour,minute;

  double A,B,C,D,E,F,alpha,day,year,factor,second;

  factor=0.5/DAY_SEC/1000; 
  F=(jd+0.5)-floor(jd+0.5);
  if ((F+factor)>=1.0) {
    jd=jd+factor;
    F=0.0;
  }

  Z=floor(jd+0.5);

  if (Z<2299161) A=Z;
  else {
    alpha=floor((Z-1867216.25)/36524.25);
    A=Z+1+alpha-floor(alpha/4);
  }

  B=A+1524;
  C=floor((B-122.1)/365.25);
  D=floor(365.25*C);
  E=floor((B-D)/30.6001);
  day=B-D-floor(30.6001*E)+F;

  if (E<13.5) month=floor(E-0.5);
  else month=floor(E-12.5);
  if (month>2.5) year=C-4716;
  else year=C-4715;
  


  *yr=(int) year;
  *mo=month;
  *dy=(int) floor(day);

  /* okay now use the residual of the day to work out the time */

  A=(day-floor(day))*DAY_SEC;

  hour=(int) (A/3600.0);
  minute=(int) ((A-hour*3600)/60);
  second=A-hour*3600-minute*60;
    
  *hr=hour;
  *mt=minute;
  *sc=second;
  return 0;
}

/**
 * @brief Convert an input time from HHMM format to number of seconds.
 *
 * @param[in] text - String with the format HHMM
 *
 * @param[out] Seconds of day corresponding to input string
 **/
double TimeStrToSOD(char *text)
{
  int i, hr, mn;

  /* Determine the location of a possible colon in the text string */
  for(i = 0; (text[i] != ':') && (text[i] != 0); i++);

  /* Only hours are specified, convert to seconds and return */
  if(text[i] == 0) return atoi(text) * 3600L;

  /* Extract both the minutes and seconds from the string */
  text[i] = 0;
  hr = atoi(text);
  mn = atoi(text + i + 1);
  return hr * 3600L + mn * 60L;
}

/**
 * @brief Convert a formatted string to an epoch time
 *
 * @param[in] text - String with the format YYYYMMDD
 *
 * @param[out] tme - Number of seconds since 00:00 UT on January 1, 1970.
 **/
double TimeStrToEpoch(char *text)
{
  double tme;
  int val;
  int yr,mo,dy;

  /* Calculate day, month, and year from YYYYMMDD format date */
  val=atoi(text);
  dy=val % 100;
  mo=(val / 100) % 100;
  yr=(val / 10000);

  /* If only 2-digit year provided then assume it was pre-2000 */
  if (yr<1970) yr+=1900;

  /* Calculate epoch time of input year, month, and day */
  tme=TimeYMDHMSToEpoch(yr,mo,dy,0,0,0);

  /* Return epoch time in number of seconds since 00:00UT on January 1, 1970 */
  return tme;
}


/*-----------------------------------------------------------------------------
;
; NAME:
;       dayno
;
; PURPOSE:
;       Function to compute the day of the year and the number of days in the
;       year.
;
; CALLING SEQUENCE:
;       doy = dayno(year, month, day, diy);
;
;     Input Arguments:
;       year          - year [1965-2014]
;       month         - month of year [01-12]
;       day           - day of month [01-31]
;
;     Output Arguments (integer pointers):
;       diy           - number of days in the given year
;
;     Return Value:
;       error code
;
;+-----------------------------------------------------------------------------
*/

int dayno(int year, int month, int day, int *diy)
{
  int k,tot;
  int ndays[] = {31,28,31,30,31,30,31,31,30,31,30,31};

  *diy = 365;
  if(((year%4==0)&&(year%100!=0))||(year%400==0)) {
    ndays[1]++;
    *diy = 366;
  }

  tot = 0;
  for (k=0; k<month-1; k++) tot += ndays[k];
  tot += day;

  return tot;
}
